test/e2e/addon_update.go

```diff
@@ -0,0 +1,247 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"time"
+
+	"github.com/tmc/scp" // scp files to the cluster
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"golang.org/x/crypto/ssh"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var addonTestPollInterval = 3 * time.Second
+var addonTestPollTimeout = 1 * time.Minute
+var addonNamespace = api.NamespaceDefault
+
+var _ = Describe("Addon update", func() {
+
+	framework := Framework{BaseName: "addon-update"}
+	var dir string
+	var sshClient *ssh.Client
+	var c *client.Client
+
+	BeforeEach(func() {
+		var err error
+		c, err = loadClient()
+		Expect(err).NotTo(HaveOccurred())
+
+		sshClient, err = getSSHClient()
+		Expect(err).NotTo(HaveOccurred())
+
+		// do not use "service" command because it clears the environment variables
+		sshExecAndVerify(sshClient, "sudo TEST_ADDON_CHECK_INTERVAL_SEC=1 /etc/init.d/kube-addons restart")
+
+		framework.beforeEach()
+		Expect(framework.Namespace).NotTo(BeNil())
+	})
+
+	AfterEach(func() {
+		if sshClient != nil {
+			// restart addon_update with the default options
+			sshExec(sshClient, "sudo /etc/init.d/kube-addons restart")
+			sshClient.Close()
+		}
+		framework.afterEach()
+	})
+
+	// WARNING: the test is not parallel-friendly!
+	It("should propagate add-on file changes", func() {
+		//these tests are long, so I squeezed several cases in one scenario
+		Expect(sshClient).NotTo(BeNil())
+
+		command := exec.Command("pwd")
+		output, err := command.CombinedOutput()
+		Expect(err).NotTo(HaveOccurred())
+
+		Logf("execute test in " + string(output))
+
+		// This test requires SSH, so the provider check should be identical to
+		// those tests.
+		if !providerIs("gce") {
+			Logf(fmt.Sprintf("Skipping test, which is not implemented for %s", testContext.Provider))
+			return
+		}
+
+		temporaryRemotePathPrefix := "addon-test-dir"
+		temporaryRemotePath := temporaryRemotePathPrefix + "/" + dir // in home directory on kubernetes-master
+		sshExecAndVerify(sshClient, fmt.Sprintf("mkdir -p %s", temporaryRemotePath))
+		defer func() {
+			// ignore the result because it's cleanup
+			sshExec(sshClient, fmt.Sprintf("rm -rf %s", temporaryRemotePathPrefix))
+		}()
+
+		// TODO: find a better way of getting to the kubernetes root dir
+		localPath := "../../../../../test/e2e/addon_update_test_manifests"
```

Comment:
Sigh, I should have caught this pattern the first time, but I was scanning with common e2e brain. We used to have a function ages ago that did exactly this. The problem is that you have to be *super* careful, because what you have right here doesn't work with with the binary test process that's running on Jenkins. We've since moved almost everything to literals in code to avoid having separate assets.

If you want to head this direction, you have to add the files to https://github.com/GoogleCloudPlatform/kubernetes/blob/0f1c4c25c344f70c3592040b2ef092ccdce0244f/hack/lib/golang.sh#L59 ... but I'm not enthralled by this. This test is really bucking the definition of e2e. :/